File: Interviews/GFG/Daily-Problems/2022/October/Day-10/Number-Of-Islands.py
from typing import List
class Solution:
    def dfs(self, r, c, n, m, grid, visited):
        if r < 0 or r >= n or c < 0 or c >= m: return
        if grid[r][c] == 0 or visited[r][c]: return
        visited[r][c] = 1
        dx, dy = [-1, 1, 0, 0], [0, 0, -1, 1]
        for i in range(4):
            self.dfs(r + dx[i], c + dy[i], n, m, grid, visited)
        
    def numOfIslands(self, n: int, m : int, operators : List[List[int]]) -> List[int]:
        k, grid = len(operators), [[0 for col in range(m)] for row in range(n)]
        ans = [0 for i in range(k)]
        
        for i in range(k):
            islands, r, c = 0, operators[i][0], operators[i][1]
            grid[r][c] = 1
            visited = [[0 for col in range(m)] for row in range(n)]
            for j in range(n):
                for k in range(m):
                    if not visited[j][k] and grid[j][k] == 1:
                        islands += 1
                        self.dfs(j, k, n, m, grid, visited)
            ans[i] = islands
        
        return ans

File: Interviews/GFG/random-problems/Medium/Number-Of-Islands/Number-Of-Islands.cpp
class Solution {
public:
    void dfs(int r, int c, int n, int m, vector<vector<int>> &grid, vector<vector<int>> &visited) {
        if (r < 0 or r >= n or c < 0 or c >= m) return;
        if (grid[r][c] == 0 or visited[r][c]) return;
        visited[r][c] = 1;
        int dx[4] = { -1, 1, 0, 0}, dy[4] = {0, 0, -1, 1};
        for (int i = 0; i < 4; i++) {
            dfs(r + dx[i], c + dy[i], n, m, grid, visited);
        }
    }
    vector<int> numOfIslands(int n, int m, vector<vector<int>> &operators) {
        vector<vector<int>> grid(n, vector<int>(m, 0));
        int k = operators.size();
        vector<int> ans(k);

        for (int i = 0; i < k; i++) {
            int islands = 0, r = operators[i][0], c = operators[i][1];
            grid[r][c] = 1;
            vector<vector<int>> visited(n, vector<int> (m, 0));
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < m; k++) {
                    if (!visited[j][k] and grid[j][k] == 1) {
                        islands += 1;
                        dfs(j, k, n, m, grid, visited);
                    }
                }
            }
            ans[i] = islands;
        }

        return ans;
    }
};